Use writer for hamiltonians

syclik opened the pull request:
#### Summary:

This propagates the writer to the hamiltonians.

#### Intended Effect:

This is a refactor of the code.
Implements more of the writer refactor in the feature/issue-1361-writer_callback branch.

#### How to Verify:

Run tests.

#### Documentation:

None.

#### Reviewer Suggestions: 

Anyone.

syclik commented on src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp:
```diff
@@ -53,38 +52,37 @@ namespace stan {
 
       virtual void update(Point& z) {
```
@betanalpha, I think the right thing is to add `base_writer` as an argument.

betanalpha replied:
That's fine.

syclik commented:
@betanalpha, can you take a look at this and make sure I didn't mess up somewhere? The stringstream creation has the same effect as in the implementation of your branch.

betanalpha commented:
Yeah... this is nothing like the writer refactor.  None of the MCMC classes should have access to the writers, instead everything was pulled out into separate classes that accepted writers as necessary; it's _much_ cleaner that way.  This branch has some bad, unnecessary interpolation of the new and the old.

syclik commented:
Hmm. You're right about the MCMC classes and writers, but I ran into the
problem of having a writer in scope, which is how I ended up here. I think
now that I'm here, I can get to where you were going with functions
accepting writers.

What else is bad and unnecessary? Just let me know and I'll try to get it
out.


On Tue, Dec 1, 2015 at 9:15 AM, Michael Betancourt <notifications@github.com
> wrote:

> Yeah... this is nothing like the writer refactor. None of the MCMC classes
> should have access to the writers, instead everything was pulled out into
> separate classes that accepted writers as necessary; it's *much* cleaner
> that way. This branch has some bad, unnecessary interpolation of the new
> and the old.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1680#issuecomment-160980651>.
>

syclik commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -1,20 +1,19 @@
     public:
-      base_mcmc(std::ostream* o, std::ostream* e)
-        : out_stream_(o), err_stream_(e) {}
+      explicit base_mcmc(interface_callbacks::writer::base_writer& writer)
```
this should be 0-arg constructor.

syclik commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -24,15 +23,15 @@ namespace stan {
       }
 
-      virtual void write_sampler_param_names(std::ostream& o) {}
+      virtual void write_sampler_param_names() {}
```
remove

syclik commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -24,15 +23,15 @@ namespace stan {
+      virtual void write_sampler_param_names() {}
 
-      virtual void write_sampler_params(std::ostream& o) {}
+      virtual void write_sampler_params() {}
```
remove

syclik commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -24,15 +23,15 @@ namespace stan {
       virtual void get_sampler_params(std::vector<double>& values) {}
 
-      virtual void write_sampler_state(std::ostream* o) {}
+      virtual void write_sampler_state() {}
```
change to 
```
virtual void get_sampler_diagnostic_names
        (std::vector<std::string>& model_names,
         std::vector<std::string>& names) {}
      virtual void get_sampler_diagnostics(std::vector<double>& values) {}
```

syclik commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -42,14 +41,10 @@ namespace stan {
-
-      std::ostream* out_stream_;
-      std::ostream* err_stream_;
+      stan::interface_callbacks::writer::base_writer& writer_;
```
remove

syclik commented on src/stan/mcmc/fixed_param_sampler.hpp:
```diff
@@ -8,14 +9,13 @@
-      explicit fixed_param_sampler(std::ostream* o,
-                                   std::ostream* e)
-        : base_mcmc(o, e) {
+      explicit fixed_param_sampler(interface_callbacks::writer::base_writer&
```
0-arg constructor

syclik commented on src/stan/mcmc/hmc/base_hmc.hpp:
```diff
@@ -23,23 +23,22 @@ namespace stan {
       base_hmc(Model &model, BaseRNG& rng,
-               std::ostream* o, std::ostream* e)
-        : base_mcmc(o, e),
+               interface_callbacks::writer::base_writer& writer)
```
remove this arg

syclik commented on src/stan/mcmc/hmc/base_hmc.hpp:
```diff
@@ -23,23 +23,22 @@ namespace stan {
-        *o << "# Step size = " << get_nominal_stepsize() << std::endl;
-        stan::interface_callbacks::writer::stream_writer writer(*o);
-        z_.write_metric(writer);
+      void write_sampler_state() {
```
add `base_writer&` as arg

syclik commented on src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp:
```diff
@@ -1,23 +1,23 @@
-      base_hamiltonian(Model& m, std::ostream* e)
-        : model_(m), err_stream_(e) {}
+      base_hamiltonian(Model& model,
+                       interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp:
```diff
@@ -53,38 +53,30 @@ namespace stan {
       virtual void update(Point& z) {
         try {
-          stan::model::gradient(model_, z.q, z.V, z.g, err_stream_);
+          stan::model::gradient(model_, z.q, z.V, z.g, writer_);
```
??? what do we do here?

syclik commented:
@betanalpha, found one of the spots that were tricky to deal with. Should
base_hamiltonian's update() method accept a writer? That seems like the
natural way to go with this.

On Tue, Dec 1, 2015 at 9:34 AM, Daniel Lee <bearlee@alum.mit.edu> wrote:

> Hmm. You're right about the MCMC classes and writers, but I ran into the
> problem of having a writer in scope, which is how I ended up here. I think
> now that I'm here, I can get to where you were going with functions
> accepting writers.
>
> What else is bad and unnecessary? Just let me know and I'll try to get it
> out.
>
>
> On Tue, Dec 1, 2015 at 9:15 AM, Michael Betancourt <
> notifications@github.com> wrote:
>
>> Yeah... this is nothing like the writer refactor. None of the MCMC
>> classes should have access to the writers, instead everything was pulled
>> out into separate classes that accepted writers as necessary; it's *much*
>> cleaner that way. This branch has some bad, unnecessary interpolation of
>> the new and the old.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/stan-dev/stan/pull/1680#issuecomment-160980651>.
>>
>
>

syclik commented on src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp:
```diff
@@ -53,38 +53,30 @@ namespace stan {
-            << "either severely ill-conditioned or misspecified."
-            << std::endl;
+      Model& model_;
+      interface_callbacks::writer::base_writer& writer_;
```
remove

syclik commented on src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp:
```diff
@@ -53,38 +53,30 @@ namespace stan {
+      Model& model_;
+      interface_callbacks::writer::base_writer& writer_;
+
+      void write_error_msg_(const std::exception& e) {
```
accept a `base_writer&` and write to it instead

syclik commented on src/stan/mcmc/hmc/hamiltonians/dense_e_point.hpp:
```diff
@@ -1,3 +1,4 @@
+
```
remove

syclik commented on src/stan/mcmc/hmc/hamiltonians/diag_e_metric.hpp:
```diff
@@ -1,22 +1,22 @@
-      diag_e_metric(Model& model, std::ostream* e)
-        : base_hamiltonian<Model, diag_e_point, BaseRNG>(model, e) {}
+      diag_e_metric(Model& model,
+                    interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/hamiltonians/unit_e_metric.hpp:
```diff
@@ -1,22 +1,23 @@
-      unit_e_metric(Model& model, std::ostream* e)
-        : base_hamiltonian<Model, unit_e_point, BaseRNG>(model, e) {}
+      unit_e_metric(Model& model,
+                    interface_callbacks::writer::base_writer& writer)
```
remote writer

syclik commented on src/stan/mcmc/hmc/integrators/base_integrator.hpp:
```diff
@@ -9,15 +9,15 @@ namespace stan {
-      explicit base_integrator(std::ostream* o)
-        : out_stream_(o) {}
+      explicit base_integrator(interface_callbacks::writer::base_writer& writer)
+        : writer_(writer) {}
```
remote writer

syclik commented on src/stan/mcmc/hmc/integrators/base_leapfrog.hpp:
```diff
@@ -11,8 +12,8 @@ namespace stan {
     public:
-      explicit base_leapfrog(std::ostream* o)
-        : base_integrator<Hamiltonian>(o) {}
+      explicit base_leapfrog(interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/integrators/base_leapfrog.hpp:
```diff
@@ -11,8 +12,8 @@ namespace stan {
+        : base_integrator<Hamiltonian>(writer) {}
 
       void evolve(typename Hamiltonian::PointType& z,
                   Hamiltonian& hamiltonian,
```
add writer

syclik commented on src/stan/mcmc/hmc/integrators/base_leapfrog.hpp:
```diff
@@ -28,31 +29,42 @@ namespace stan {
       void verbose_evolve(typename Hamiltonian::PointType& z,
                           Hamiltonian& hamiltonian,
                           const double epsilon) {
```
add writer

syclik commented on src/stan/mcmc/hmc/integrators/expl_leapfrog.hpp:
```diff
@@ -10,8 +11,8 @@ namespace stan {
     public:
-      explicit expl_leapfrog(std::ostream* o = 0)
-        : base_leapfrog<Hamiltonian>(o) {}
+      explicit expl_leapfrog(interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/nuts/adapt_dense_e_nuts.hpp:
```diff
@@ -16,8 +17,8 @@ namespace stan {
         adapt_dense_e_nuts(Model &model, BaseRNG& rng,
-                           std::ostream* o, std::ostream* e)
-          : dense_e_nuts<Model, BaseRNG>(model, rng, o, e),
+                           interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp:
```diff
@@ -17,8 +18,8 @@ namespace stan {
         adapt_diag_e_nuts(Model &model, BaseRNG& rng,
-                          std::ostream* o, std::ostream* e)
-          : diag_e_nuts<Model, BaseRNG>(model, rng, o, e),
+                          interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp:
```diff
@@ -17,8 +18,8 @@ namespace stan {
       adapt_unit_e_nuts(Model &model, BaseRNG& rng,
-                        std::ostream* o, std::ostream* e)
-        : unit_e_nuts<Model, BaseRNG>(model, rng, o, e) {}
+                        interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/nuts/base_nuts.hpp:
```diff
@@ -32,8 +33,8 @@ namespace stan {
       base_nuts(Model &model, BaseRNG& rng,
-                std::ostream* o, std::ostream* e)
-        : base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(model, rng, o, e),
+                interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/nuts/dense_e_nuts.hpp:
```diff
@@ -15,10 +16,10 @@ namespace stan {
-      dense_e_nuts(Model &model, BaseRNG& rng, std::ostream* o,
-                   std::ostream* e)
+      dense_e_nuts(Model &model, BaseRNG& rng,
+                   interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp:
```diff
@@ -17,9 +18,8 @@ namespace stan {
-                               std::ostream* o,
-                               std::ostream* e)
-        : dense_e_static_hmc<Model, BaseRNG>(model, rng, o, e),
+                               interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp:
```diff
@@ -16,11 +16,10 @@ namespace stan {
-          : diag_e_static_hmc<Model, BaseRNG>(model, rng, o, e),
-          stepsize_var_adapter(model.num_params_r()) {}
+      adapt_diag_e_static_hmc(Model &model, BaseRNG& rng,
+                              interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp:
```diff
@@ -17,8 +17,8 @@ namespace stan {
       adapt_unit_e_static_hmc(Model &model, BaseRNG& rng,
-                              std::ostream* o, std::ostream* e)
-        : unit_e_static_hmc<Model, BaseRNG>(model, rng, o, e) { }
+                              interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/base_static_hmc.hpp:
```diff
@@ -23,8 +23,8 @@ namespace stan {
       base_static_hmc(Model &model, BaseRNG& rng,
-                      std::ostream* o, std::ostream* e)
-        : base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(model, rng, o, e),
+                      interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/dense_e_static_hmc.hpp:
```diff
@@ -19,9 +20,9 @@ namespace stan {
     public:
       dense_e_static_hmc(Model &model, BaseRNG& rng,
-                         std::ostream* o, std::ostream* e)
+                         stan::interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/diag_e_static_hmc.hpp:
```diff
@@ -19,9 +20,9 @@ namespace stan {
     public:
       diag_e_static_hmc(Model &model, BaseRNG& rng,
-                        std::ostream* o, std::ostream* e)
+                        interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/mcmc/hmc/static/unit_e_static_hmc.hpp:
```diff
@@ -18,9 +19,9 @@ namespace stan {
     public:
       unit_e_static_hmc(Model &model, BaseRNG& rng,
-                        std::ostream* o, std::ostream* e)
+                        interface_callbacks::writer::base_writer& writer)
```
remove writer

syclik commented on src/stan/model/util.hpp:
```diff
@@ -418,6 +419,22 @@ namespace stan {
     }
 
     template <class M>
+    void gradient(const M& model,
```
note to self: is this necessary now?

syclik commented:
@betanalpha, I'll update the pull request. I'll let you know when I'm done.

betanalpha commented:
Yes.  We had an entire thread about exactly this point a few weeks ago.

I gave the base_hamiltonian class a string buffer that could store error
messages and then write them out to, say, a writer but we decided that
it was best to simply pass a writer to the update call.


On Dec 1, 2015, at 2:44 PM, Daniel Lee <notifications@github.com> wrote:

> @betanalpha, found one of the spots that were tricky to deal with. Should
> base_hamiltonian's update() method accept a writer? That seems like the
> natural way to go with this.
> 
> On Tue, Dec 1, 2015 at 9:34 AM, Daniel Lee <bearlee@alum.mit.edu> wrote:
> 
> > Hmm. You're right about the MCMC classes and writers, but I ran into the
> > problem of having a writer in scope, which is how I ended up here. I think
> > now that I'm here, I can get to where you were going with functions
> > accepting writers.
> >
> > What else is bad and unnecessary? Just let me know and I'll try to get it
> > out.
> >
> >
> > On Tue, Dec 1, 2015 at 9:15 AM, Michael Betancourt <
> > notifications@github.com> wrote:
> >
> >> Yeah... this is nothing like the writer refactor. None of the MCMC
> >> classes should have access to the writers, instead everything was pulled
> >> out into separate classes that accepted writers as necessary; it's *much*
> >> cleaner that way. This branch has some bad, unnecessary interpolation of
> >> the new and the old.
> >>
> >> —
> >> Reply to this email directly or view it on GitHub
> >> <https://github.com/stan-dev/stan/pull/1680#issuecomment-160980651>.
> >>
> >
> >
> —
> Reply to this email directly or view it on GitHub.
>

betanalpha commented:
Make sure you look through base_mcmc and base_hmc on
https://github.com/stan-dev/stan/tree/feature/issue-1361-writer_callback.
Some of the methods that used the internal streams were removed as they 
were not used anywhere.

On Dec 1, 2015, at 2:34 PM, Daniel Lee <notifications@github.com> wrote:

> Hmm. You're right about the MCMC classes and writers, but I ran into the
> problem of having a writer in scope, which is how I ended up here. I think
> now that I'm here, I can get to where you were going with functions
> accepting writers.
> 
> What else is bad and unnecessary? Just let me know and I'll try to get it
> out.
> 
> 
> On Tue, Dec 1, 2015 at 9:15 AM, Michael Betancourt <notifications@github.com
> > wrote:
> 
> > Yeah... this is nothing like the writer refactor. None of the MCMC classes
> > should have access to the writers, instead everything was pulled out into
> > separate classes that accepted writers as necessary; it's *much* cleaner
> > that way. This branch has some bad, unnecessary interpolation of the new
> > and the old.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/stan-dev/stan/pull/1680#issuecomment-160980651>.
> >
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
@betanalpha: I went through and added a `stan::interface_callbacks::writer::base_writer` argument to `base_mcmc.hpp` and propagated it through all places that touched it. See b4019ab. (I haven't removed the member variables.

The reason I implemented it the way I did was because by adding `base_writer` to arguments of functions, it seems inconsistent when some class methods have writers and others don't. I guess this could be an indication that some of the class methods has the side effect of generating output while others don't. If that's the intention, I'm ok with it. This isn't true of all the classes. I think it's mostly true of the Hamiltonians. 

In the Hamiltonians, these functions have to have `base_writer`:
- `init`
- `update`

and these don't:
- `T`
- `V`
- `tau`
- `phi`
- `H`
- `dtau_dq`
- `dtau_dp`
- `dphi_dq`
- `sample_p`

I think I'd feel a little more sane if all those functions required `base_writer` or none or it was clear why something like `sample_p` doesn't need one.

And here are the source files that changed (I excluded tests). These all had to add `base_writer` to an argument to make this fly. I think this was why I just started carrying around a member variable:
- `src/stan/mcmc/base_mcmc.hpp`
- `src/stan/mcmc/fixed_param_sampler.hpp`
- `src/stan/mcmc/hmc/base_hmc.hpp`
- `src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp`
- `src/stan/mcmc/hmc/integrators/base_integrator.hpp`
- `src/stan/mcmc/hmc/integrators/base_leapfrog.hpp`
- `src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp`
- `src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp`
- `src/stan/mcmc/hmc/nuts/base_nuts.hpp`
- `src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp`
- `src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp`
- `src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp`
- `src/stan/mcmc/hmc/static/base_static_hmc.hpp`
- `src/stan/services/mcmc/sample.hpp`
- `src/stan/services/mcmc/warmup.hpp`
- `src/stan/services/sample/generate_transitions.hpp`
- `src/stan/services/sample/init_adapt.hpp`

betanalpha commented:
This is so confusing — you keep changing your perspective on this.  You added
everything despite the refactor not doing that, then yesterday you said that you’d
remove them.  What’s going on?

Yes, only a few of the methods have the side effect of producing output and the refactor
isolated the writer to those methods, removing all internal references.  The only other
method in the entire MCMC hierarchy that should need a writer is transition, which
would be defined as virtual in base_mcmc and then implemented in the various HMC
classes.

On Dec 2, 2015, at 9:35 PM, Daniel Lee <notifications@github.com> wrote:

> @betanalpha: I went through and added a stan::interface_callbacks::writer::base_writer argument to base_mcmc.hpp and propagated it through all places that touched it. See b4019ab. (I haven't removed the member variables.
> 
> The reason I implemented it the way I did was because by adding base_writer to arguments of functions, it seems inconsistent when some class methods have writers and others don't. I guess this could be an indication that some of the class methods has the side effect of generating output while others don't. If that's the intention, I'm ok with it. This isn't true of all the classes. I think it's mostly true of the Hamiltonians.
> 
> In the Hamiltonians, these functions have to have base_writer:
> 
> init
> update
> and these don't:
> 
> T
> V
> tau
> phi
> H
> dtau_dq
> dtau_dp
> dphi_dq
> sample_p
> I think I'd feel a little more sane if all those functions required base_writer or none or it was clear why something like sample_p doesn't need one.
> 
> And here are the source files that changed (I excluded tests). These all had to add base_writer to an argument to make this fly. I think this was why I just started carrying around a member variable:
> 
> src/stan/mcmc/base_mcmc.hpp
> src/stan/mcmc/fixed_param_sampler.hpp
> src/stan/mcmc/hmc/base_hmc.hpp
> src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp
> src/stan/mcmc/hmc/integrators/base_integrator.hpp
> src/stan/mcmc/hmc/integrators/base_leapfrog.hpp
> src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp
> src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp
> src/stan/mcmc/hmc/nuts/base_nuts.hpp
> src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp
> src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp
> src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp
> src/stan/mcmc/hmc/static/base_static_hmc.hpp
> src/stan/services/mcmc/sample.hpp
> src/stan/services/mcmc/warmup.hpp
> src/stan/services/sample/generate_transitions.hpp
> src/stan/services/sample/init_adapt.hpp
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
I didn't change my perspective. That single commit was adding the writer to every place that needed it. I haven't removed the writers from all the constructors. (In other words, the pull request isn't finished.)

I agree with you that transition needs it, but you're excluding a lot of other methods. The diff should be the minimal set of methods that need it. For example:
- base_hmc: init_stepsize -> hamiltonian's init -> hamiltonian's update
- base_leapfrog: evolve -> hamiltonian's update. This forces base_integrator to change it's update method which in turn forces all the integrators to update.
- base_nuts: transition -> hamiltonian's update. This forces all of the subclasses of base_nuts to update.
- base_nuts: build_tree -> integrator's evolve.
and then there were a few more.

I think the few methods propagate the writer to a lot of other places. Am I mistaken?

I'll remove the member variables next. I just got this compiling and passing all tests, so I committed for us to look at the diff.

betanalpha commented:
You know I can’t read these diffs — I need to see the final implementation as I’m not
sure where you’re going with all of this until you present a final PR.

On Dec 2, 2015, at 10:02 PM, Daniel Lee <notifications@github.com> wrote:

> I didn't change my perspective. That single commit was adding the writer to every place that needed it. I haven't removed the writers from all the constructors. (In other words, the pull request isn't finished.)
> 
> I agree with you that transition needs it, but you're excluding a lot of other methods. The diff should be the minimal set of methods that need it. For example:
> 
> base_hmc: init_stepsize -> hamiltonian's init -> hamiltonian's update
> base_leapfrog: evolve -> hamiltonian's update. This forces base_integrator to change it's update method which in turn forces all the integrators to update.
> base_nuts: transition -> hamiltonian's update. This forces all of the subclasses of base_nuts to update.
> base_nuts: build_tree -> integrator's evolve. and then there were a few more.
> I think the few methods propagate the writer to a lot of other places. Am I mistaken?
> 
> I'll remove the member variables next. I just got this compiling and passing all tests, so I committed for us to look at the diff.
> 
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
Fair point. It'll probably take me another day or two, but if you're feeling proactive:
https://github.com/stan-dev/stan/commit/b4019abbdc7bddfbde7a12f27208d83ca498f030

syclik commented:
(there's a link in my last comment... GitHub shrunk it)

bob-carpenter commented:
I don't think it should be in the base_mcmc class.

Writers are stateful, so you have to ask yourself if
the state of the writer is part of a base_mcmc object.

I'd think probably not and therefore you'd want to include
them as arguments where needed.

Basically, anything that reduces statefulness of classes
is a good thing because it's easier to reason about programs.

- Bob

> On Dec 2, 2015, at 4:35 PM, Daniel Lee <notifications@github.com> wrote:
> 
> @betanalpha: I went through and added a stan::interface_callbacks::writer::base_writer argument to base_mcmc.hpp and propagated it through all places that touched it. See b4019ab. (I haven't removed the member variables.
> 
> The reason I implemented it the way I did was because by adding base_writer to arguments of functions, it seems inconsistent when some class methods have writers and others don't. I guess this could be an indication that some of the class methods has the side effect of generating output while others don't. If that's the intention, I'm ok with it. This isn't true of all the classes. I think it's mostly true of the Hamiltonians.
> 
> In the Hamiltonians, these functions have to have base_writer:
> 
> 	• init
> 	• update
> and these don't:
> 
> 	• T
> 	• V
> 	• tau
> 	• phi
> 	• H
> 	• dtau_dq
> 	• dtau_dp
> 	• dphi_dq
> 	• sample_p
> I think I'd feel a little more sane if all those functions required base_writer or none or it was clear why something like sample_p doesn't need one.
> 
> And here are the source files that changed (I excluded tests). These all had to add base_writer to an argument to make this fly. I think this was why I just started carrying around a member variable:
> 
> 	• src/stan/mcmc/base_mcmc.hpp
> 	• src/stan/mcmc/fixed_param_sampler.hpp
> 	• src/stan/mcmc/hmc/base_hmc.hpp
> 	• src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp
> 	• src/stan/mcmc/hmc/integrators/base_integrator.hpp
> 	• src/stan/mcmc/hmc/integrators/base_leapfrog.hpp
> 	• src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp
> 	• src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp
> 	• src/stan/mcmc/hmc/nuts/base_nuts.hpp
> 	• src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp
> 	• src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp
> 	• src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp
> 	• src/stan/mcmc/hmc/static/base_static_hmc.hpp
> 	• src/stan/services/mcmc/sample.hpp
> 	• src/stan/services/mcmc/warmup.hpp
> 	• src/stan/services/sample/generate_transitions.hpp
> 	• src/stan/services/sample/init_adapt.hpp
> —
> Reply to this email directly or view it on GitHub.
>

betanalpha commented:
I think we're all on the same page, Dan's just making incremental commits.

> On Dec 2, 2015, at 22:57, Bob Carpenter <notifications@github.com> wrote:
> 
> I don't think it should be in the base_mcmc class.
> 
> Writers are stateful, so you have to ask yourself if
> the state of the writer is part of a base_mcmc object.
> 
> I'd think probably not and therefore you'd want to include
> them as arguments where needed.
> 
> Basically, anything that reduces statefulness of classes
> is a good thing because it's easier to reason about programs.
> 
> - Bob
> 
> > On Dec 2, 2015, at 4:35 PM, Daniel Lee <notifications@github.com> wrote:
> > 
> > @betanalpha: I went through and added a stan::interface_callbacks::writer::base_writer argument to base_mcmc.hpp and propagated it through all places that touched it. See b4019ab. (I haven't removed the member variables.
> > 
> > The reason I implemented it the way I did was because by adding base_writer to arguments of functions, it seems inconsistent when some class methods have writers and others don't. I guess this could be an indication that some of the class methods has the side effect of generating output while others don't. If that's the intention, I'm ok with it. This isn't true of all the classes. I think it's mostly true of the Hamiltonians.
> > 
> > In the Hamiltonians, these functions have to have base_writer:
> > 
> > • init
> > • update
> > and these don't:
> > 
> > • T
> > • V
> > • tau
> > • phi
> > • H
> > • dtau_dq
> > • dtau_dp
> > • dphi_dq
> > • sample_p
> > I think I'd feel a little more sane if all those functions required base_writer or none or it was clear why something like sample_p doesn't need one.
> > 
> > And here are the source files that changed (I excluded tests). These all had to add base_writer to an argument to make this fly. I think this was why I just started carrying around a member variable:
> > 
> > • src/stan/mcmc/base_mcmc.hpp
> > • src/stan/mcmc/fixed_param_sampler.hpp
> > • src/stan/mcmc/hmc/base_hmc.hpp
> > • src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp
> > • src/stan/mcmc/hmc/integrators/base_integrator.hpp
> > • src/stan/mcmc/hmc/integrators/base_leapfrog.hpp
> > • src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp
> > • src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp
> > • src/stan/mcmc/hmc/nuts/base_nuts.hpp
> > • src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp
> > • src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp
> > • src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp
> > • src/stan/mcmc/hmc/static/base_static_hmc.hpp
> > • src/stan/services/mcmc/sample.hpp
> > • src/stan/services/mcmc/warmup.hpp
> > • src/stan/services/sample/generate_transitions.hpp
> > • src/stan/services/sample/init_adapt.hpp
> > —
> > Reply to this email directly or view it on GitHub.
> > 
> 
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
Yup. We're all on the same page. It was just a struggle getting it plumbed through everything, but now we're good. I just have a bit more to do for this pull request. 

> On Dec 2, 2015, at 5:59 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> I think we're all on the same page, Dan's just making incremental commits.
> 
> > On Dec 2, 2015, at 22:57, Bob Carpenter <notifications@github.com> wrote:
> > 
> > I don't think it should be in the base_mcmc class.
> > 
> > Writers are stateful, so you have to ask yourself if
> > the state of the writer is part of a base_mcmc object.
> > 
> > I'd think probably not and therefore you'd want to include
> > them as arguments where needed.
> > 
> > Basically, anything that reduces statefulness of classes
> > is a good thing because it's easier to reason about programs.
> > 
> > - Bob
> > 
> > > On Dec 2, 2015, at 4:35 PM, Daniel Lee <notifications@github.com> wrote:
> > > 
> > > @betanalpha: I went through and added a stan::interface_callbacks::writer::base_writer argument to base_mcmc.hpp and propagated it through all places that touched it. See b4019ab. (I haven't removed the member variables.
> > > 
> > > The reason I implemented it the way I did was because by adding base_writer to arguments of functions, it seems inconsistent when some class methods have writers and others don't. I guess this could be an indication that some of the class methods has the side effect of generating output while others don't. If that's the intention, I'm ok with it. This isn't true of all the classes. I think it's mostly true of the Hamiltonians.
> > > 
> > > In the Hamiltonians, these functions have to have base_writer:
> > > 
> > > • init
> > > • update
> > > and these don't:
> > > 
> > > • T
> > > • V
> > > • tau
> > > • phi
> > > • H
> > > • dtau_dq
> > > • dtau_dp
> > > • dphi_dq
> > > • sample_p
> > > I think I'd feel a little more sane if all those functions required base_writer or none or it was clear why something like sample_p doesn't need one.
> > > 
> > > And here are the source files that changed (I excluded tests). These all had to add base_writer to an argument to make this fly. I think this was why I just started carrying around a member variable:
> > > 
> > > • src/stan/mcmc/base_mcmc.hpp
> > > • src/stan/mcmc/fixed_param_sampler.hpp
> > > • src/stan/mcmc/hmc/base_hmc.hpp
> > > • src/stan/mcmc/hmc/hamiltonians/base_hamiltonian.hpp
> > > • src/stan/mcmc/hmc/integrators/base_integrator.hpp
> > > • src/stan/mcmc/hmc/integrators/base_leapfrog.hpp
> > > • src/stan/mcmc/hmc/nuts/adapt_diag_e_nuts.hpp
> > > • src/stan/mcmc/hmc/nuts/adapt_unit_e_nuts.hpp
> > > • src/stan/mcmc/hmc/nuts/base_nuts.hpp
> > > • src/stan/mcmc/hmc/static/adapt_dense_e_static_hmc.hpp
> > > • src/stan/mcmc/hmc/static/adapt_diag_e_static_hmc.hpp
> > > • src/stan/mcmc/hmc/static/adapt_unit_e_static_hmc.hpp
> > > • src/stan/mcmc/hmc/static/base_static_hmc.hpp
> > > • src/stan/services/mcmc/sample.hpp
> > > • src/stan/services/mcmc/warmup.hpp
> > > • src/stan/services/sample/generate_transitions.hpp
> > > • src/stan/services/sample/init_adapt.hpp
> > > —
> > > Reply to this email directly or view it on GitHub.
> > > 
> > 
> > —
> > Reply to this email directly or view it on GitHub.
> > 
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
Jenkins, retest this please.

syclik commented:
Jenkins, retest this please.

syclik commented:
Jenkins, retest this please.

syclik commented:
@betanalpha, this should be fixed now. Let me know if you spot anything off.

syclik commented:
Jenkins, retest this please.

betanalpha commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -1,38 +1,42 @@
-      virtual void write_sampler_params(std::ostream& o) {}
+      virtual void
+      write_sampler_params(interface_callbacks::writer::base_writer& writer) {}
 
```
"write_sampler_param_names" and "write_sampler_params" should not be used anywhere and can be deleted.

betanalpha commented on src/stan/mcmc/base_mcmc.hpp:
```diff
@@ -42,14 +46,8 @@ namespace stan {
 
     protected:
       std::string name_;
```
We should define name_ on construction and have name get passed through all of the MCMC derivations to satisfy the "make all of the things defined on construction" idiom, not how I originally did it.  If you want to avoid this here then I'd make it an issue.

syclik replied:
Do we ever make use of `name_` or `name()`? I couldn't find any uses of either.

betanalpha replied:
It's used in stansummary.

> On Dec 8, 2015, at 20:23, Daniel Lee <notifications@github.com> wrote:
> 
> In src/stan/mcmc/base_mcmc.hpp:
> 
> > @@ -42,14 +46,8 @@ namespace stan {
> >  
> >      protected:
> >        std::string name_;
> Do we ever make use of name_ or name()? I couldn't find any uses of either.
> 
> —
> Reply to this email directly or view it on GitHub.
>

betanalpha commented:
There are two functions in base_mcmc to delete but otherwise :+1:.  Just need to make all of the chances to services/sample now...

syclik commented:
Thank you. I knew having another set of eyes would be good. I'll make those
changes and merge when it's finished testing.

On Tue, Dec 8, 2015 at 5:13 PM, Michael Betancourt <notifications@github.com
> wrote:

> There are two functions in base_mcmc to delete but otherwise [image: :+1:].
> Just need to make all of the chances to services/sample now...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1680#issuecomment-163034888>.
>

syclik commented:
@betanalpha, the last commit removes uses of `name_`. If you think it's worth keeping around, I can revert that last commit.

betanalpha commented:
See my last email, this should break cmdstan.

> On Dec 8, 2015, at 21:01, Daniel Lee <notifications@github.com> wrote:
> 
> @betanalpha, the last commit removes uses of name_. If you think it's worth keeping around, I can revert that last commit.
> 
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
I don't see it used anywhere in CmdStan directly. If it doesn't get triggered, think it'll be safe to remove?

betanalpha commented:
You’re right — stansummary uses the arguments to reconstruct the sampler name.
I guess this must be an artifact of some previous use.  It’s useful information but
I would not protest strongly against its removal given that it would be easy to add
back in.

On Dec 8, 2015, at 9:19 PM, Daniel Lee <notifications@github.com> wrote:

> I don't see it used anywhere in CmdStan directly. If it doesn't get triggered, think it'll be safe to remove?
> 
> —
> Reply to this email directly or view it on GitHub.
>

syclik commented:
Cool. I think the tests should pass now.

On Tue, Dec 8, 2015 at 9:36 PM, Michael Betancourt <notifications@github.com
> wrote:

> You’re right — stansummary uses the arguments to reconstruct the sampler
> name.
> I guess this must be an artifact of some previous use. It’s useful
> information but
> I would not protest strongly against its removal given that it would be
> easy to add
> back in.
>
> On Dec 8, 2015, at 9:19 PM, Daniel Lee <notifications@github.com> wrote:
>
> > I don't see it used anywhere in CmdStan directly. If it doesn't get
> triggered, think it'll be safe to remove?
>
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/stan/pull/1680#issuecomment-163087869>.
>
